### PR TITLE
Correct Handling of some Mistypings and Oversize Values

### PR DESCRIPTION
**Pull request by rmarrowstone**
* Over-precise and oversize Decimals are now errors
* Can coerce IonInt to Decimal (given the above)
* Over-precise Timestamps are now errors (whether to Trino TImestamp or Date)
* Correctly Typed null TLVs are now handled as empty structs
* Fixed Exception thrown for mistyping cases and added test cases.

### CODE REVIEW

File: lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java [hunk 1]
```diff
@@ -79,6 +79,44 @@ public void testSuperBasicStruct()
...
+                "{ foo: null.list, bar: false }",
+                "{ foo: 12345,     bar: null.int }");
+
+        for (String ion : ions) {
+            Assertions.assertThrows(TrinoException.class, () -> {
+                assertValues(rowType, ion, List.of());
+            });
+        }
+    }
+
```
Comment (linlin-s) [inline lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java:119]: I noticed that we already have a [PR](https://github.com/rmarrowstone/trino/pull/6) that provides setups for mistyping for each type. I was wondering if we might consider using those setups and then create test cases for each types, rather than adding the `testColumnMistypings` in this PR?

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonDecoderFactory.java [hunk 2]
```diff
@@ -248,23 +250,30 @@ private static BlockDecoder timestampDecoder(TimestampType type)
...
-                        .unscaledValue());
-                type.writeObject(builder, unscaled);
-            };
-        }
+                        .unscaledValue();
+
+                if (Decimals.overflows(unscaled, type.getPrecision())) {
+                    throw new NumberFormatException(
+                            "Decimal value %s does not fit %d digits of precision and %d of scale!"
+                                    .formatted(reader.bigDecimalValue(), type.getPrecision(), type.getScale()));
```
Comment (linlin-s) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonDecoderFactory.java:267]: Nit pick: We might consider creating local variables for `precision` and `scale` which could be reused throughout method.

File: lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java [hunk 3]
```diff
@@ -204,17 +204,53 @@ public void testDecimalPrecisionAndScale()
...
+    @Test
+    public void testNumbersTooBigForDecimal128()
+    {
+        RowType rowType = RowType.rowType(
+                field("amount", DecimalType.createDecimalType(20, 2)));
+
+        List<String> ions = List.of(
+                "{ amount: 12345678901234567890.4 }",
+                "{ amount: 1.234 }",
```
Comment (linlin-s) [inline lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java:324]: Nit: Would it be more explicit to modify the test case from `1.234` to `12345678901234567890.123?` My understanding is that we're testing numbers too large for Decimal128, specifically where the scale overflows. To make this scenario more explicit, should we ensure that the test case itself is a Decimal128 with an overflow scale?

File: lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java (hunk 1 above)
Comment (rmarrowstone) [reply lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java:119]: Yea... I thought we had decided we weren't going to have a committed A/B test setup, so I thought we were going to scrap that. I'll follow up with desaikd on that.

File: lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java (hunk 3 above)
Comment (rmarrowstone) [reply lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java:324]: That is a different than this but yea that would be a good one to prove.

**Comment by rmarrowstone** [comment]
I'm going to close this to redo the Decimal and Timestamp behavior and split out the different parts.
